### PR TITLE
Zip packager helper

### PR DESCRIPTION
**Pull request by yichen88**
Signed-off-by: yichen88 <tang.yichenyves@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
feature
Add more functionality for packager a zip bytes 

**What is the current behavior?** *(You can also link to an open issue here)*
Only zip files.

**What is the new behavior (if this is a feature change)?**
Could add String or bytes in zip bytes.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [x] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

### CODE REVIEW

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        workingDir = fileSystem.getPath("/wd");
+        Path f1 = workingDir.resolve("f1");
+        Path f2 = workingDir.resolve("f2");
+        Path f3 = workingDir.resolve("f3.gz");
+        try {
+            Files.createDirectories(workingDir);
+            try (BufferedWriter f1Writer = Files.newBufferedWriter(f1, StandardCharsets.UTF_8);
+                 BufferedWriter f2Writer = Files.newBufferedWriter(f2, StandardCharsets.UTF_8);
+                 OutputStream f3Writer = new GZIPOutputStream(Files.newOutputStream(f3))) {
+                f1Writer.write("foo");
```
Comment (sylvlecl) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: Maybe add an accent or another non-ASCII character in one of the files.

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                .addString("k2", "é")
+                .addString("k3", "î", StandardCharsets.ISO_8859_1)
+                .addBytes("nested.zip", nestedZipBytes);
+        byte[] zipBytes = zipPackager.toZipBytes();
+
+        IOUtils.copy(new ByteArrayInputStream(zipBytes), Files.newOutputStream(workingDir.resolve("res.zip")));
+
+        ZipFile zipFile = new ZipFile(workingDir.resolve("res.zip"));
+        InputStream f1 = zipFile.getInputStream("f1");
+        assertEquals("foo", IOUtils.toString(zipFile.getInputStream("f1"), StandardCharsets.UTF_8));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: sonar : zipFile.getInputStream("f1") might be null --> replace with : Objects.requireNonNull(zipFile.getInputStream("f1")

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("bar", IOUtils.toString(zipFile.getInputStream("f2"), StandardCharsets.UTF_8));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("hello", IOUtils.toString(zipFile.getInputStream("f3"), StandardCharsets.UTF_8));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("v1", IOUtils.toString(zipFile.getInputStream("k1"), StandardCharsets.UTF_8));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("é", IOUtils.toString(zipFile.getInputStream("k2"), StandardCharsets.UTF_8));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("î", IOUtils.toString(zipFile.getInputStream("k3"), StandardCharsets.ISO_8859_1));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

File: commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        assertArrayEquals(nestedZipBytes, IOUtils.toByteArray(zipFile.getInputStream("nested.zip")));
```
Comment (ChamseddineBhd) [inline commons/src/test/java/com/powsybl/commons/compress/ZipPackagerTest.java]: same here

**Comment by sylvlecl** [comment]
> 
> 
> Shouldn't we add a static metod : public static byte[] archiveFilesToZipBytes(Path baseDir, List fileNames) to avoid object instantiation in some cases ?

Maybe we should keep also the deprecated class, which provided such methods ?
To decrease duplication, I think it should re-use the new ZipPackager class.

**Comment by yichen88** [comment]
> avoid object instantiation

but why should we avoid object instantiation? This instance is not used a lot.

**Comment by ChamseddineBhd** [comment]
> not

If we want just to just to get the array of bytes, it's more interesting to call  just the static method public `static byte[] archiveFilesToZipBytes(Path baseDir, List fileNames)`
`ZipHelper.archiveFilesToZipBytes(path,fileNames)`  than 
` ZipPackager zip = new ZipPackager();
zip.addPath(path);
zip.toZipBytes();`

**Comment by ChamseddineBhd** [comment]
> > avoid object instantiation
> 
> but why should we avoid object instantiation? This instance is not used a lot.

If we want just to just to get the array of bytes, it's more interesting to call  just the static method public `static byte[] archiveFilesToZipBytes(Path baseDir, List fileNames)`
`ZipHelper.archiveFilesToZipBytes(path,fileNames)`  than 
` ZipPackager zip = new ZipPackager();
zip.addPath(path);
zip.toZipBytes();`

**Comment by ChamseddineBhd** [comment]
> > Shouldn't we add a static metod : public static byte[] archiveFilesToZipBytes(Path baseDir, List fileNames) to avoid object instantiation in some cases ?
> 
> Maybe we should keep also the deprecated class, which provided such methods ?
> To decrease duplication, I think it should re-use the new ZipPackager class.

we can even implement the method in the zipPackager class.

**Comment by sylvlecl** [comment]
> 
> we can even implement the method in the zipPackager class.

Personally I would prefer not to mix the utility methods with the object itself. It's just a matter of taste, I just find it a little less confusing, and I think it's the most used convention in the JDK/guava etc. The `ZipPackager` sounds more like something to be used as an object :)

Not blocking at all if we keep it in `ZipPackager` though.

**Pull request merged**